router/dataplane.go

```diff
@@ -2651,49 +2841,103 @@ func (p *scionPacketProcessor) currentHopPointer( /*@ ghost ubScionL []byte @*/
-// @ requires  p.EqAbsHopField(oldPkt)
-// @ requires  p.EqAbsInfoField(oldPkt)
-// @ ensures   reserr == nil ==> AbsVerifyCurrentMACConstraint(oldPkt, dp)
+// @ requires  slayers.ValidPktMetaHdr(ubScionL) && p.scionLayer.EqAbsHeader(ubScionL)
+// @ requires  absPkt(ubScionL).PathNotFullyTraversed()
+// @ requires  p.EqAbsHopField(absPkt(ubScionL))
+// @ requires  p.EqAbsInfoField(absPkt(ubScionL))
+// @ ensures   reserr == nil ==> slayers.ValidPktMetaHdr(ubScionL) && p.scionLayer.EqAbsHeader(ubScionL)
+// @ ensures   reserr == nil ==> absPkt(ubScionL).PathNotFullyTraversed()
+// @ ensures   reserr == nil ==> AbsVerifyCurrentMACConstraint(absPkt(ubScionL), dp)
+// @ ensures   reserr == nil ==> old(slayers.IsSupportedPkt(ubScionL)) == slayers.IsSupportedPkt(ubScionL)
+// @ ensures   reserr == nil ==> absPkt(ubScionL) == old(absPkt(ubScionL))
+// @ ensures   reserr == nil ==> p.DstIsLocalIngressID(ubScionL) == old(p.DstIsLocalIngressID(ubScionL))
+// @ ensures   reserr == nil ==> p.LastHopLen(ubScionL) == old(p.LastHopLen(ubScionL))
 // @ ensures   reserr != nil && respr.OutPkt != nil ==>
 // @ 	absIO_val(respr.OutPkt, respr.EgressID).isIO_val_Unsupported
 // @ decreases
-func (p *scionPacketProcessor) verifyCurrentMAC( /*@ ghost oldPkt io.IO_pkt2, ghost dp io.DataPlaneSpec @*/ ) (respr processResult, reserr error) {
+func (p *scionPacketProcessor) verifyCurrentMAC( /*@ ghost dp io.DataPlaneSpec, ghost ubScionL []byte, ghost ubLL []byte, ghost startLL int, ghost endLL int @*/ ) (respr processResult, reserr error) {
+	// @ ghost oldPkt := absPkt(ubScionL)
 	fullMac := path.FullMAC(p.mac, p.infoField, p.hopField, p.macBuffers.scionInput)
 	// @ fold acc(sl.Bytes(p.hopField.Mac[:path.MacLen], 0, path.MacLen), R21)
 	// @ defer unfold acc(sl.Bytes(p.hopField.Mac[:path.MacLen], 0, path.MacLen), R21)
 	// @ sl.SplitRange_Bytes(fullMac, 0, path.MacLen, R21)
 	// @ ghost defer sl.CombineRange_Bytes(fullMac, 0, path.MacLen, R21)
 	if subtle.ConstantTimeCompare(p.hopField.Mac[:path.MacLen], fullMac[:path.MacLen]) == 0 {
-		// @ ToDoAfterScionFix("https://github.com/scionproto/scion/issues/4482")
-		return p.packSCMP(
+		// @ ghost ubPath := p.scionLayer.UBPath(ubScionL)
+		// @ ghost start := p.scionLayer.PathStartIdx(ubScionL)
+		// @ ghost end   := p.scionLayer.PathEndIdx(ubScionL)
+		// @ assert ubScionL[start:end] === ubPath
+
+		// @ unfold acc(p.scionLayer.Mem(ubScionL), R13)
+		// @ defer fold acc(p.scionLayer.Mem(ubScionL), R13)
+		// @ unfold acc(p.path.Mem(ubPath), R14)
+		// @ defer fold acc(p.path.Mem(ubPath), R14)
+		// @ unfold acc(p.path.Base.Mem(), R15)
+		// @ defer fold acc(p.path.Base.Mem(), R15)
+		// (VerifiedSCION) Gobra is unable to verify permissions for `p.path.PathMeta.CurrINF`
+		// and `p.path.PathMeta.CurrHF` directly at the function call.
+		// To work around this limitation, these fields are first stored in temporary variables.
+		tmpCurrINF := p.path.PathMeta.CurrINF
+		tmpCurrHF := p.path.PathMeta.CurrHF
```

Comment:
I still find this supper puzzling. Can you please remind me if you were using any kind of `unfolding` expression in the function call?

Reply:
I tried all possible combinations of `unfold` and `unfolding`. As soon as the field access is directly a function argument, Gobra complains about missing permissions.

Reply:
Did you get this error even without any unfolds/unfoldings?

Reply:
Yes, even without Gobra complains about missing permissions.